components/engine_rocks/src/misc.rs

```diff
@@ -1,9 +1,67 @@
 // Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
 
 use crate::engine::RocksEngine;
+use crate::import::RocksIngestExternalFileOptions;
+use crate::sst::RocksSstWriterBuilder;
 use crate::util;
-use engine_traits::{CFNamesExt, MiscExt, Range, Result, ALL_CFS};
+use engine_traits::{
+    CFHandleExt, CFNamesExt, DeleteStrategy, ImportExt, IngestExternalFileOptions, IterOptions,
+    Iterable, Iterator, MiscExt, Mutable, Range, Result, SstWriter, SstWriterBuilder,
+    WriteBatchExt, WriteOptions, ALL_CFS,
+};
 use rocksdb::Range as RocksRange;
+use tikv_util::keybuilder::KeyBuilder;
+
+pub const MAX_DELETE_COUNT_BY_KEY: usize = 2048;
+
+impl RocksEngine {
+    // We store all data which would be deleted in memory at first because the data of region will never be larger than
+    // max-region-size.
+    fn delete_all_in_range_cf_by_ingest(
+        &self,
+        cf: &str,
+        opts: IterOptions,
+        sst_path: String,
+        start_key: &[u8],
+    ) -> Result<usize> {
+        let mut data: Vec<Vec<u8>> = vec![];
+        let mut it = self.iterator_cf_opt(cf, opts)?;
+        let mut it_valid = it.seek(start_key.into())?;
+        while it_valid {
+            data.push(it.key().to_vec());
+            it_valid = it.next()?;
+        }
+        let count = data.len();
+        if data.len() > MAX_DELETE_COUNT_BY_KEY {
+            let builder = RocksSstWriterBuilder::new().set_db(self).set_cf(cf);
+            let mut writer = builder.build(sst_path.as_str())?;
+            for key in data.iter() {
+                writer.delete(key).unwrap();
+            }
+            writer.finish()?;
+            let handle = self.cf_handle(cf)?;
+            let mut opt = RocksIngestExternalFileOptions::new();
+            opt.move_files(true);
+            self.ingest_external_file_cf(handle, &opt, &[sst_path.as_str()])?;
+        } else {
+            let mut wb = self.write_batch();
+            let mut opt = WriteOptions::new();
+            opt.set_sync(false);
+            for key in data.iter() {
+                wb.delete_cf(cf, key)?;
+                if wb.count() >= Self::WRITE_BATCH_MAX_KEYS {
+                    self.write_opt(&wb, &opt)?;
+                    wb.clear();
+                }
+            }
+            if wb.count() > 0 {
+                self.write_opt(&wb, &opt)?;
+            }
+            self.sync_wal()?;
+        }
+        Ok(count)
+    }
+}
 
 impl MiscExt for RocksEngine {
     fn is_titan(&self) -> bool {
```

Comment:
we can remove it from trait now

Reply:
Why?

Reply:
after moving the logic to `engine_rocks`, it only used in `engine_rocks`